weaviate/connect/connection.py

```diff
@@ -67,94 +73,115 @@ def __init__(
+                    # credentials should only be saved for client credentials, otherwise use refresh token
+                    self._create_background_token_refresh(_auth)
+                else:
+                    self._create_background_token_refresh()
             else:
                 raise AuthenticationFailedException(
                     f""""No login credentials provided. The weaviate instance at {self.url} requires login credential,
                     use argument 'auth_client_secret'."""
                 )
-        elif response.status_code == 404 and self._auth_client_secret is not None:
+        elif response.status_code == 404 and auth_client_secret is not None:
             _Warnings.auth_with_anon_weaviate()
+            self._session = requests.Session()
+        else:
+            self._session = requests.Session()
 
-    def __del__(self):
-        """
-        Destructor for Connection class instance.
-        """
+    def _create_background_token_refresh(self, _auth: Optional[_Auth] = None):
+        """Create a background thread that periodically refreshes access and refresh tokens.
 
-        self._session.close()
+        While the underlying library refreshes tokens, it does not have an internal cronjob that checks every
+        X-seconds if a token has expired. If there is no activity for longer than the refresh tokens lifetime, it will
+        expire. Therefore, refresh manually shortly before expiration time is up."""
+        if "refresh_token" not in self._session.token and _auth is None:
+            return
 
-    def _refresh_authentication(self) -> None:
-        """Refreshes the bearer token if current authentication is expired
+        expires_in: int = self._session.token.get(
+            "expires_in", 60
+        )  # use 1minute as token lifetime if not supplied
+
+        def periodic_refresh_token(refresh_time: int, _auth: Optional[_Auth]):
+            while True:
+                time.sleep(min(refresh_time - 5, 1))
+                # use refresh token when available
+                if "refresh_token" in self._session.token:
+                    self._session.token = self._session.refresh_token(
+                        self._session.metadata["token_endpoint"]
+                    )
+                    refresh_time = self._session.token.get("expires_in")
+                else:
+                    # client credentials usually does not contain a refresh token => get a new token using the
+                    # saved credentials
+                    assert _auth is not None
+                    new_session = _auth.get_auth_session()
+                    self._session.token = new_session.fetch_token()
+
+        self._background_thread = Thread(
+            target=periodic_refresh_token,
+            args=(expires_in, _auth),
+            daemon=True,
+            name="TokenRefresh",
+        )
+        self._background_thread.start()
```

Comment:
is this thread automatically cleaned up in case of an ungraceful client crash?

Reply:
good point, done